### ISSUE TITLE
Remove overlap between dotLess.Core and dotLess.ClientOnly

### ISSUE THREAD

**Issue by Daniel15** [dotless/dotless#304]
Imagine a scenario like this:
Third-party library A references dotLess.Core
Third-party library B references dotLess.ClientOnly

In this scenario, libraries A and B can't be used together in the same project, as dotLess.Core and dotLess.ClientOnly overlap in terms of namespaces and classes. Ideally this overlap should be removed, and dotLess.Core should have a dependency on dotLess.ClientOnly, just adding the additional functionality on top of it.

Also see andrewdavey/cassette#379

**Comment by WalkerCodeRanger** [comment]
I too think this should be cleaned up. One package should reference the other.  However, I would add that while that change is being made, The package names should be cleaned up/reworked.  Core implies to me that it should be what is in ClientOnly, but a better name than Core could be chosen.  The name ClientOnly is totally opaque to me and implies nothing useful.  Also, I see no reason not to ship the command line compiler with the equivalent of the ClientOnly package.

Perhaps things should be further factored out, issue #328 suggests adding build targets which might make sense as its own package.

**Comment by rytmis** [comment]
The naming is probably a source of much confusion. dotless.Core.dll and dotless.ClientOnly.dll are basically the same thing, with the difference that dotless.Core.dll has dotless.AspNet.dll ILMerge'd in (both assemblies have Pandora, ServiceLocation and whatnot ILMerged too).

We already have separate projects for Core, Compiler and AspNet -- we just do away with the distinction as a part of the build process. In the future, the package layout might reflect that:

dotless.Core - compiler infrastructure only
dotless.Commandline/dotless.Compiler - command line compiler tools
dotless.AspNet - equivalent to the current dotless package

And possibly dotless.Owin for projects that target ASP.NET 5 and the like.

**Comment by Daniel15** [comment]
>  with the difference that dotless.Core.dll has dotless.AspNet.dll ILMerge'd

Why do they need to be ILMerge'd? Seems like it'd be simpler if a NuGet `dotless.AspNet` package **only** included `dotless.AspNet.dll`, and had a dependency on the "core" package to pull in the core stuff.

> And possibly dotless.Owin for projects that target ASP.NET 5 and the like.

For what it's worth, ASP.NET 5 doesn't use OWIN. Its design was inspired by OWIN and middleware you write for OWIN can be used with minimal changes (eg. see https://github.com/reactjs/React.NET/blob/master/src/React.AspNet/JsxFileSystem.cs and https://github.com/reactjs/React.NET/blob/master/src/React.AspNet/JsxFileMiddleware.cs) but they're two separate libraries.

**Comment by rytmis** [comment]
> >  with the difference that dotless.Core.dll has dotless.AspNet.dll ILMerge'd

>Why do they need to be ILMerge'd? Seems like it'd be simpler if a NuGet `dotless.AspNet` package 
>**only** included `dotless.AspNet.dll`, and had a dependency on the "core" package to pull in the core 
>stuff.

You'll have to ask the earlier maintainers for that -- I wasn't around for that decision. :(

> > And possibly dotless.Owin for projects that target ASP.NET 5 and the like.
> For what it's worth, ASP.NET 5 doesn't use OWIN.

Strictly speaking that's 100% correct -- http://owin.org/ lists ASP.NET vNext as an _OWIN implementation_. ;-)

**Comment by Tigraine** [comment]
The build process and this decision was made when there was no NuGet yet.
At the time I was a firm believer that it's more convenient to users to
have just one .dll to be referenced from inside their project. And any
project that had dotless as it's dependency was usually shipping it as a
.dll alongside their own so you would not get any issues.
In case the project came with the ClientOnly dll it also worked with the
Core anyways.
So we internalized all the stuff that's internal (Pandora Container was
used and written mostly because nobody uses it and it's essentially the
internal IoC of dotless exclusively) to make it more convenient and not
scare users by a ton of dlls that all have to be referenced and then
clutter their namespace.

Cleaning this up ( and the naming ) should be quite simple. And I'm all for
it. Now that probably 99% of the users just install the dotless NuGet we
can remove this.

greetings DAniel

Lauri Kotilainen <notifications@github.com> schrieb am Fr., 15. Mai 2015 um
19:24 Uhr:

**Comment by superguppie** [comment]
...nobody uses it...

Until now.

I need to inject parameters. You suggested at https://groups.google.com/forum/embed/#!topic/dotless/Q2_YQSKdC2c to provide an IParameterSource.
The infrastructure for it is there. All I need to do is create an Engine that takes a ContainerFactory instance that registers my IParameterSource.
Except. I can't make a ContainerFactory that registers anything, due to the Pandora parts being inaccessible.

I can see where the decissions leading to this come from. I'm sympathetic to it.
But I do think it is a shame that something was made that can be extended so nicely, only to break it by making an essential part internal.

When is the cleanup going to be?

Until then I will resort to getting the code and including it in my project.

**Comment by rytmis** [comment]
Yeah, I know, the current situation sucks. For what it's worth, you can use the ParameterDecorator implementation of ILessEngine to inject an IParameterSource to a wrapped ILessEngine -- that's how  QueryStringParameterSource is used, too.

As for the cleanup -- I don't have a timeframe for you, unfortunately. I had high hopes for working on dotless this autumn, but between work and house renovation, coming up with the energy for that has been a bit of a challenge.

Version-wise, I'd say 2.0 is where it is most likely to happen, as that's the perfect time for breaking changes. I was planning on a roadmap that only includes one minor version (1.6) prior to 2.0.

**Comment by superguppie** [comment]
Errr....QueryStringParameterSource/AspNetContainerFactory also depends on access to Pandora.
Not going to subtype ParameterDecorator to tweak it. Good code needs to be left in peace. And considering that ILessEngines also need to be registered using Pandora, I am sceptical about getting that working anyway.

I am actually using a ContainerFactory subtype and an IParameterResource implementation just like QueryStringParameterSource/AspNetContainerFactory do,
I just had to pull the source for the dotless.Core project into my own solution, and put the Pandora dll in it separately. No code changes needed. Just making Pandora public. Then it worked like a charm. Just shows how good the code already is.

I'll keep an eye out for updates. Until then, the current solution is good enough.
